sklearn/datasets/_base.py

```diff
@@ -1454,28 +1464,160 @@ def _fetch_remote(remote, dirname=None, n_retries=3, delay=1):
+            raise OSError(
+                f"The SHA256 checksum of {remote.filename} ({checksum}) "
+                f"differs from expected ({remote.checksum})."
+            )
+    except BaseException:
+        os.unlink(temp_file.name)
+        raise
+
+    # The following renaming is atomic whenever temp_file_path and
+    # file_path are on the same filesystem. This should be the case most of
+    # the time, but we still use shutil.move instead of os.rename in case
+    # they are not.
+    shutil.move(temp_file_path, file_path)
 
-    file_path = remote.filename if dirname is None else join(dirname, remote.filename)
-    while True:
-        try:
-            urlretrieve(remote.url, file_path)
-            break
-        except (URLError, TimeoutError):
-            if n_retries == 0:
-                # If no more retries are left, re-raise the caught exception.
-                raise
-            warnings.warn(f"Retry downloading from url: {remote.url}")
-            n_retries -= 1
-            time.sleep(delay)
-
-    checksum = _sha256(file_path)
-    if remote.checksum != checksum:
-        raise OSError(
-            "{} has an SHA256 checksum ({}) "
-            "differing from expected ({}), "
-            "file may be corrupted.".format(file_path, checksum, remote.checksum)
-        )
     return file_path
+
+
+def _filter_filename(value, filter_dots=True):
+    """Derive a name that is safe to use as filename from the given string.
+
+    Adapted from the `slugify` function of django:
+    https://github.com/django/django/blob/master/django/utils/text.py
+
+    Convert spaces or repeated dashes to single dashes. Replace characters that
+    aren't alphanumerics, underscores, hyphens or dots by underscores. Convert
+    to lowercase. Also strip leading and trailing whitespace, dashes, and
+    underscores.
+    """
+    value = unicodedata.normalize("NFKD", value).lower()
+    if filter_dots:
+        value = re.sub(r"[^\w\s-]+", "_", value)
+    else:
+        value = re.sub(r"[^.\w\s-]+", "_", value)
+    value = re.sub(r"[\s-]+", "-", value)
+    return value.strip("-_.")
+
+
+def _derive_folder_and_filename_from_url(url):
+    parsed_url = urlparse(url)
+    if not parsed_url.hostname:
+        raise ValueError(f"Invalid URL: {url}")
+    folder_components = [_filter_filename(parsed_url.hostname, filter_dots=False)]
+    path = parsed_url.path
+
+    if "/" in path:
+        base_folder, raw_filename = path.rsplit("/", 1)
+
+        base_folder = _filter_filename(base_folder)
+        if base_folder:
+            folder_components.append(base_folder)
+    else:
+        raw_filename = path
+
+    filename = _filter_filename(raw_filename, filter_dots=False)
+    if not filename:
+        filename = "downloaded_file"
+
+    return "/".join(folder_components), filename
+
+
+def fetch_file(
+    url, folder=None, local_filename=None, sha256=None, n_retries=3, delay=1
+):
```

Comment:
Is the naming discussion still on ?

If so I'd like to propose `fetch_dataset`. It keeps the same semantic as other fetchers. Other fetchers fetch a specific dataset, fetch_openml fetches datasets from openml and this one can fetch any dataset from anywhere.

If not, you can ignore and merge :smile:

Reply:
I prefer `fetch_file`. While this function would primarily be used to fetch CSV or parquet files, nothing prevents using it do download any other kind of files (e.g. pre-trained machine learning models saved using the skops format for instance).

Reply:
I have to say I was not a fan originally of the name, but I have grown accustomed to `fetch_file` with time spent reviewing this PR ...

I guess if some people feel really strongly about it, there is a bit of time to change it before the 1.6 release :wink:

Reply:
> nothing prevents using it do download any other kind of files

it's in the datastet module though :)

Reply:
Alright, it was just in case you were still hesitating.